[WIP] Common test for equivalence between sparse and dense matrices.

wdevazelhes opened the pull request:
Fixes #1572 
Follow-up of #7590 
TODO:
- [x] <strike>Try to be more robust by maybe testing only the score (if possible), with a certain tolerance, and for certain random seeds (or if testing the predict also maybe test that like 90% of predictions are the same ?)</strike> 
After discussion with @agramfort , it would be better to test on a realistic dataset which would avoid weird instable behaviours
- [x] address https://github.com/scikit-learn/scikit-learn/pull/13246#pullrequestreview-208581402
- [x] Remove the fix for the Affinity propagation and put it in another PR
- [ ] Wait for the tests to pass thanks to other PR's fix (see for instance<strike> #13336 </strike> #17554 )
- [x] Replace the sparse format generator in ``check_estimator_sparse_dense`` by the one (more exhaustive) in ``check_estimator_sparse_data``
- [x] Address @jnothman's comment https://github.com/scikit-learn/scikit-learn/pull/13246#discussion_r532265287, and be more exhaustive in the fake dataset creation, in particular ensuring it has some zeros (for the sparsity to actually do sth), and <strike> that it tests well multiclass and multioutputs estimators </strike> <b>->see if it's necessary here to check with multioutput sparse datasets</b>
- [ ] Look at sparse matrices which may contain duplicates (see one of the comments in the thread I think)

maniteja123 committed:
Add test for sparse and dense equivalence

maniteja123 committed:
Set random state

Commit:
Merge branch 'master' into tests_sparse

# Conflicts:
#	sklearn/utils/estimator_checks.py

Commit:
wip tests for sparse estimators

Commit:
remove wrongly merged code

Commit:
Fix: make it work

wdevazelhes commented:
I tried to integrate the dictionary of values to overwrite the default, as in this @lesteve comment: https://github.com/scikit-learn/scikit-learn/pull/7590#issuecomment-260640615
My strategy is to look at tests that fail, and if they fail because of different parameters settings than default, I would add them to this dictionary, otherwise I'll file the bug. Does this sound good @jnothman, @lesteve ?

wdevazelhes commented:
Right now here are the tests that fail (without for now looking at whether it's a real fail, or if it's because the input takes different paths between sparse and non sparse): 
- TransformedTargetRegressor
- SGDRegressor
- SGDClassifier
- RobustScaler
- RidgeClassifierCV
- RidgeCV
- Ridge
- RANSACRegressor
- Perceptron
- PassiveAggressiveRegressor
- PassiveAggressiveClassifier
- MultiOutputRegressor
- LinearRegression
- KNeighborsRegressor
- KNeighborsClassifier
- GradientBoostingClassifier

agramfort commented:
for linear models it's likely that sparse data are not centered to output
will differ.
only coordinate_descent will implicitly center sparse data.

>

wdevazelhes commented:
Setting `fit_intercept=False`, these are the tests that fail:
TransformedTargetRegressor
RobustScaler
RANSACRegressor
MultiOutputRegressor
MultiOutputRegressor
KNeighborsRegressor
KNeighborsClassifier
GradientBoostingClassifier

Commit:
Add fit_intercept=False for linear models

Commit:
add brute exception for kneighbors classifier

agramfort committed:
more workaround

agramfort committed:
fix test by fixing the seed :)

Commit:
STY: fix pep8 errors

Commit:
Merge branch 'tests_sparse' of https://github.com/wdevazelhes/scikit-learn into tests_sparse

Commit:
FIX deal with the case that was failing (test_check_estimator_pairwise)

wdevazelhes commented on sklearn/neighbors/regression.py:
```diff
@@ -148,7 +148,7 @@ def predict(self, X):
         """
         if issparse(X) and self.metric == 'precomputed':
-            raise ValueError(
+            raise TypeError(
```
@agramfort should  I leave this for another PR or is it ok to change it here ?

wdevazelhes replied:
@agramfort related to our discussion, in fact here they do call `check_array` but the sparse error is raised by this "if", which has a `ValueError`, so I guess this is the right fix (and I'll let it in this PR for this one as you said)

wdevazelhes commented on sklearn/utils/estimator_checks.py:
```diff
@@ -2406,7 +2406,7 @@ def check_estimator_sparse_dense(name, estimator_orig):
                 assert_equal(probs.shape, (X.shape[0], 4))
         except TypeError as e:
-            if 'sparse' not in repr(e):
+            if 'sparse' not in str.lower(repr(e)):
```
I put this because the estimator that used kernel='precomputed' raised something like "Sparse..." which is an OK message

Commit:
FIX fix random seed to make test_check_estimator pass (lines at the end of the test)

wdevazelhes commented on sklearn/utils/estimator_checks.py:
```diff
@@ -2338,7 +2338,7 @@ def check_classifiers_regression_target(name, estimator_orig):
 @ignore_warnings(category=(DeprecationWarning, FutureWarning))
 def check_estimator_sparse_dense(name, estimator_orig):
-    rng = np.random.RandomState(42)
+    rng = np.random.RandomState(52)
```
@agramfort I had the same problem of random seed here for `AdaBoostClassifier`: (tested at the end of `test_check_estimator`) 
this random seed fixed it

Commit:
Fix the test that was checking a ValueError

Commit:
FIX: do another way around to pass tests for kernel algorithms (because same n_samples as n_features created other bugs, and fix pep8 errors

Commit:
FIX: fix mistake on n_samples, n_features

Commit:
remove spurious comment

Commit:
Also convert X_csr

Commit:
Convert X_sp instead

Commit:
Fix typo

wdevazelhes commented:
The new seed (52) introduced a weird bug for AffinityPropagation that predicts two really different arrays (one with zeros everywhere and one with -1 everywhere (putting back the seed to 0 fixes the pb). I'll try to investigate that

Err message: 
```pytb
Testing started at 09:57 ...
/home/will/anaconda3/envs/sprint/bin/python /snap/pycharm-community/112/helpers/pycharm/_jb_pytest_runner.py --target test_common.py::test_estimators -- -k test_estimators[AffinityPropagation-check_estimator_sparse_dense]
Launching pytest with arguments -k test_estimators[AffinityPropagation-check_estimator_sparse_dense] test_common.py::test_estimators in /home/will/Code/sklearn-forks/wdevazelhes/scikit-learn/sklearn/tests

============================= test session starts ==============================
platform linux -- Python 3.7.2, pytest-4.3.0, py-1.7.0, pluggy-0.8.1
rootdir: /home/will/Code/sklearn-forks/wdevazelhes/scikit-learn, inifile: setup.cfg
plugins: sugar-0.9.2collected 5385 items / 5384 deselected / 1 selected

test_common.py FEstimator AffinityPropagation doesn't seem to fail gracefully on sparse data: it should raise a TypeError if sparse input is explicitly not supported.

sklearn/tests/test_common.py:101 (test_estimators[AffinityPropagation-check_estimator_sparse_dense])
estimator = AffinityPropagation(affinity='euclidean', convergence_iter=15, copy=True,
                    damping=0.5, max_iter=5, preference=None, verbose=False)
check = <function check_estimator_sparse_dense at 0x7fa5ee03e048>

    @pytest.mark.parametrize(
            "estimator, check",
            _generate_checks_per_estimator(_yield_all_checks,
                                           _tested_estimators()),
            ids=_rename_partial
    )
    def test_estimators(estimator, check):
        # Common tests for estimator instances
        with ignore_warnings(category=(DeprecationWarning, ConvergenceWarning,
                                       UserWarning, FutureWarning)):
            set_checking_parameters(estimator)
            name = estimator.__class__.__name__
>           check(name, estimator)

test_common.py:114: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../utils/testing.py:350: in wrapper
    return fn(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

name = 'AffinityPropagation'
estimator_orig = AffinityPropagation(affinity='euclidean', convergence_iter=15, copy=True,
                    damping=0.5, max_iter=5, preference=None, verbose=False)

    @ignore_warnings(category=(DeprecationWarning, FutureWarning))
    def check_estimator_sparse_dense(name, estimator_orig):
        rng = np.random.RandomState(52)
        X = rng.rand(40, 10)
        X[X < .8] = 0
        X_csr = sparse.csr_matrix(X)
        y = (4 * rng.rand(40)).astype(np.int)
        estimator = clone(estimator_orig)
        estimator_sp = clone(estimator_orig)
        for sparse_format in ['csr', 'csc', 'dok', 'lil', 'coo', 'dia', 'bsr']:
            X_sp = X_csr.asformat(sparse_format)
            # catch deprecation warnings
            with ignore_warnings(category=DeprecationWarning):
                if name in ['Scaler', 'StandardScaler']:
                    estimator.set_params(with_mean=False)
                    estimator_sp.set_params(with_mean=False)
            dense_vs_sparse_additional_params = defaultdict(dict,
                    {'Ridge': {'solver': 'cholesky'}})
            params = dense_vs_sparse_additional_params[
                estimator.__class__.__name__]
            estimator.set_params(**params)
            estimator_sp.set_params(**params)
            set_random_state(estimator)
            set_random_state(estimator_sp)
            try:
                with ignore_warnings(category=DeprecationWarning):
                    estimator_sp.fit(X_sp, y)
                    estimator.fit(X, y)
                if hasattr(estimator, "predict"):
                    pred = estimator.predict(X)
                    pred_sp = estimator_sp.predict(X_sp)
>                   assert_array_almost_equal(pred, pred_sp, 2)
E                   AssertionError: 
E                   Arrays are not almost equal to 2 decimals
E                   
E                   (mismatch 100.0%)
E                    x: array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
E                          0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])
E                    y: array([-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
E                          -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
E                          -1, -1, -1, -1, -1, -1])

../utils/estimator_checks.py:2367: AssertionError
                                                         [100%]

=================================== FAILURES ===================================
______ test_estimators[AffinityPropagation-check_estimator_sparse_dense] _______

estimator = AffinityPropagation(affinity='euclidean', convergence_iter=15, copy=True,
                    damping=0.5, max_iter=5, preference=None, verbose=False)
check = <function check_estimator_sparse_dense at 0x7fa5ee03e048>

    @pytest.mark.parametrize(
            "estimator, check",
            _generate_checks_per_estimator(_yield_all_checks,
                                           _tested_estimators()),
            ids=_rename_partial
    )
    def test_estimators(estimator, check):
        # Common tests for estimator instances
        with ignore_warnings(category=(DeprecationWarning, ConvergenceWarning,
                                       UserWarning, FutureWarning)):
            set_checking_parameters(estimator)
            name = estimator.__class__.__name__
>           check(name, estimator)

test_common.py:114: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../utils/testing.py:350: in wrapper
    return fn(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

name = 'AffinityPropagation'
estimator_orig = AffinityPropagation(affinity='euclidean', convergence_iter=15, copy=True,
                    damping=0.5, max_iter=5, preference=None, verbose=False)

    @ignore_warnings(category=(DeprecationWarning, FutureWarning))
    def check_estimator_sparse_dense(name, estimator_orig):
        rng = np.random.RandomState(52)
        X = rng.rand(40, 10)
        X[X < .8] = 0
        X_csr = sparse.csr_matrix(X)
        y = (4 * rng.rand(40)).astype(np.int)
        estimator = clone(estimator_orig)
        estimator_sp = clone(estimator_orig)
        for sparse_format in ['csr', 'csc', 'dok', 'lil', 'coo', 'dia', 'bsr']:
            X_sp = X_csr.asformat(sparse_format)
            # catch deprecation warnings
            with ignore_warnings(category=DeprecationWarning):
                if name in ['Scaler', 'StandardScaler']:
                    estimator.set_params(with_mean=False)
                    estimator_sp.set_params(with_mean=False)
            dense_vs_sparse_additional_params = defaultdict(dict,
                    {'Ridge': {'solver': 'cholesky'}})
            params = dense_vs_sparse_additional_params[
                estimator.__class__.__name__]
            estimator.set_params(**params)
            estimator_sp.set_params(**params)
            set_random_state(estimator)
            set_random_state(estimator_sp)
            try:
                with ignore_warnings(category=DeprecationWarning):
                    estimator_sp.fit(X_sp, y)
                    estimator.fit(X, y)
                if hasattr(estimator, "predict"):
                    pred = estimator.predict(X)
                    pred_sp = estimator_sp.predict(X_sp)
>                   assert_array_almost_equal(pred, pred_sp, 2)
E                   AssertionError: 
E                   Arrays are not almost equal to 2 decimals
E                   
E                   (mismatch 100.0%)
E                    x: array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
E                          0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])
E                    y: array([-1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
E                          -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
E                          -1, -1, -1, -1, -1, -1])
```

Commit:
Merge branch 'master' into tests_sparse

Commit:
FIX: fix teh check that was making fail AffinityPropagation and remove the fit_intercept=False that should not be needed since #13279 is merged

wdevazelhes commented:
I tried to remove the forcing of "fit_intercept=False", and it seems that some linear estimators are still bugging: here are the ones that bug:
         - PassiveAggressiveClassifier
         - PassiveAggressiveRegressor
         - Perceptron
         - Ridge
         - RidgeCV
         - RidgeClassifier
         - RidgeClassifierCV
         - SGDClassifier
         - SGDRegressor

wdevazelhes commented on sklearn/cluster/affinity_propagation_.py:
```diff
@@ -403,7 +403,7 @@ def predict(self, X):
                              "affinity='precomputed'.")
 
-        if self.cluster_centers_.size > 0:
+        if self.cluster_centers_.shape[0] > 0:
```
in fact there was a bug with the random seed I had the cluster centers were [[0, 0..0]] (unique zero cluster center), and in the sparse version it was the same but in a sparse representation hence doing size on it returned 0

wdevazelhes replied:
This sensitivity to the random seed made me think that maybe I should improve the test by testing several random seeds and only asserting that the result is close to a certain tolerance ?

wdevazelhes replied:
I just opened a PR here #13334 to isolate the change, I'll remove it from this PR

jeremiedbb commented on sklearn/utils/estimator_checks.py:
```diff
@@ -2359,6 +2362,89 @@ def check_classifiers_regression_target(name, estimator_orig):
 
+@ignore_warnings(category=(DeprecationWarning, FutureWarning))
+def check_estimator_sparse_dense(name, estimator_orig):
+    rng = np.random.RandomState(52)
```
Finding the right seed to make the test pass doesn't feel right at all :/

wdevazelhes replied:
I agree :p that's what we did with @agramfort and after discussion with @ogrisel because GradientBoosting worked but some times and some times not. But that's also why I commented https://github.com/scikit-learn/scikit-learn/pull/13246#discussion_r260758475 and added this to the todo: 

```
- [ ] Try to be more robust by maybe testing only the score (if possible), with a certain tolerance, and for certain random seeds (or if testing the predict also maybe test that like 90% of predictions are the same ?)
```

Does this sounds a good strategy ?

(Testing for a lot of random seeds, but with a certain tolerance. And for predictions since they are hard values test something like 95% of the values are the same or sth like that)

wdevazelhes replied:
Eventually I put a friendly dataset as discussed with @gramfort, so now it solves the problem

ogrisel replied:
This is bad to have checks that depend on a particular seed. Those checks are not meant to be used only by scikit-learn but also by third party libraries. So we need to make sure that the check is generic enough to avoid brittle failures (unless its caused by a limitation of a particular estimator in which case it should be XFAILED with an estimator tag and a proper motivation).

> Eventually I put a friendly dataset as discussed with @gramfort, so now it solves the problem

So this mean that this check would pass for all scikit-learn estimators with a canonical random seed such as 0 or 42? If so please change the seed to 0 or 42 because 52 makes us think that it's a special seed.

jeremiedbb commented on sklearn/utils/estimator_checks.py:
```diff
@@ -2359,6 +2362,89 @@ def check_classifiers_regression_target(name, estimator_orig):
+                pred = estimator.predict(X)
+                pred_sp = estimator_sp.predict(X_sp)
+                assert_array_almost_equal(pred, pred_sp, 2)
+                assert_equal(pred.shape, pred_sp.shape)
```
It's better to use `assert_all_close`
and `assert ... == ...`

wdevazelhes replied:
Thanks, will do

jeremiedbb commented on sklearn/utils/estimator_checks.py:
```diff
@@ -2359,6 +2362,89 @@ def check_classifiers_regression_target(name, estimator_orig):
+                    probs = probs[0]
+
+                assert_equal(probs.shape, (X.shape[0], 4))
+        except TypeError as e:
```
assert ... == ...

wdevazelhes replied:
Thanks, will do

jeremiedbb reviewed (commented):
Just a few comments

wdevazelhes commented:
> Just a few comments

Thanks for the review @jeremiedbb !

Commit:
FIX: remove the change that belongs to #13334

Commit:
Merge branch 'master' into tests_sparse

Commit:
Add more friendly dataset to avoid problems of random seeds with random data

Commit:
Address jeremie's review

wdevazelhes commented:
I addressed your comments @jeremiedbb

Commit:
FIX: put friendly dataset that allows to clean up nearest neighbors

wdevazelhes commented:
@agramfort I was able to clean up nearest neighbors with a nicer dataset (because it has no duplicates) in the last commit

so I guess it's just about the rest of the estimator to deal with now

wdevazelhes commented:
However, even if RidgeRegression is solved, there still remains
- PassiveAggressiveClassifier
- PassiveAggressiveRegressor
- Perceptron
- SGDClassifier
- SGDRegressor

(which all basically call SGD code)
Should I try to solve SGD the same way as the regressions ?
I am not familiar with cython though (but eager to learn), I don't know how hard the fix would be in cython

wdevazelhes commented:
Maybe related to https://github.com/scikit-learn/scikit-learn/issues/13367

agramfort committed:
simplify

agramfort committed:
Merge branch 'master' into tests_sparse

agramfort commented:
@wdevazelhes I pushed a fix

basically if you had random 0. in X you make features noisy and unrelated to y.
this causes a pb for SGD code that uses different heuristics in the sparse case
to decrease the learning rates for the intercept.

let's see if this turns green...

Commit:
Merge branch 'master' into tests_sparse

# Conflicts:
#	sklearn/utils/estimator_checks.py

Commit:
Merge branch 'tests_sparse' of https://github.com/wdevazelhes/scikit-learn into tests_sparse

Commit:
Merge branch 'master' into tests_sparse

jnothman commented:
> let's see if this turns green...

Unfortunately it didn't, @agramfort.

If "SGD code uses different heuristics in the sparse case to decrease the learning rates for the intercept" should we expect it to pass these tests, or be an exception?

agramfort committed:
fix precomputed case + SGD

agramfort committed:
Merge branch 'master' into tests_sparse

agramfort committed:
more try

agramfort commented:
I gave it one more try but I really need the bug fixed in the Ridge class. I skipped SGD estimators as they use different numerics in sparse vs dense case.

any help appreciated :)

jeromedockes committed:
Merge remote-tracking branch 'upstream/master' into common_check_sparse_dense_updated

jeromedockes committed:
only 2 classes in check_estimator_sparse_dense if tag "binary_only"

jeromedockes committed:
check value of multioutput or multioutput_only in tags

jeromedockes committed:
Merge remote-tracking branch 'upstream/master' into common_check_sparse_dense_updated

agramfort commented:
this one is finally green with the help of @jeromedockes and @wdevazelhes 

reviews welcome

rth commented on sklearn/utils/estimator_checks.py:
```diff
@@ -2384,6 +2385,71 @@ def check_classifiers_regression_target(name, estimator_orig):
+    centers = 2 if tags["binary_only"] else None
+    X, y = make_blobs(random_state=rng, cluster_std=0.5, centers=centers)
+    # for put some points to zero to have a little bit of sparsity
+    X -= np.min(X) - 1.  # we need positive data for ComplementNB for instance
```
This could/should use the `requires_positive_X` estimator tag.

wdevazelhes replied:
Agreed, I did the replacement now, although it will work only when #14680 is merged

rth commented on sklearn/utils/estimator_checks.py:
```diff
@@ -2384,6 +2385,71 @@ def check_classifiers_regression_target(name, estimator_orig):
+
+                if isinstance(probs, list):
+                    # XXX : hack for dummy classifier
+                    probs = probs[0]
```
Coverage says it's never run, are you sure it's needed?

wdevazelhes replied:
I don't remember why we needed it, I removed it, we can see if tests pass ?

rth commented on sklearn/utils/estimator_checks.py:
```diff
@@ -2384,6 +2385,71 @@ def check_classifiers_regression_target(name, estimator_orig):
+                assert probs.shape == (X.shape[0], len(np.unique(y)))
+        except TypeError as e:
+            if 'sparse' not in str.lower(repr(e)):
+                print("Estimator %s doesn't seem to fail gracefully on "
```
These and the below cases are not covered by tests. I think we should create a tests with a simple estimator to check this branch.

Also print goes to stdout while raised exceptions are shown on stderr (I think). That could lead to subtitle issues, but I'm not sure if it's worth addressing.

wdevazelhes replied:
I modified an already existing test to cover this (see comments below)

Regarding the print, there is some pre-existing explanation in the test (copied below):
```
    # the check for sparse input handling prints to the stdout,
    # instead of raising an error, so as not to remove the original traceback.
    # that means we need to jump through some hoops to catch it.
```

rth reviewed (commented):
Thanks @wdevazelhes ! Looks nice overall. Please remove `sklearn/linear_model/tests/__pycache__/tmp1ax1d_im`.

Commit:
Address https://github.com/scikit-learn/scikit-learn/pull/13246/files#r309167624

Commit:
Remove file that should not be here

wdevazelhes commented on sklearn/utils/tests/test_estimator_checks.py:
```diff
@@ -152,16 +152,29 @@ def fit(self, X, y):
 class NoSparseClassifier(BaseBadClassifier):
     def fit(self, X, y):
-        X, y = check_X_y(X, y, accept_sparse=['csr', 'csc'])
+        X, y = check_X_y(X, y, accept_sparse=['csr', 'csc', 'dok', 'lil',
```
I extended `NoSparseClassifier` to accept all kind of sparse (we want to test the error messages in all kind of sparsity no?), and I made it return an Exception instead (wrong type of error), and I created another one (`NoSparseClassifierTypeError`) that returns a TypeError (good type of error) but with the wrong message.

Then I replaced the pre-existing test in `test_check_estimator` that just tested the first part of the message (`Estimator %s doesn't seem to fail gracefully on sparse data`), by two tests that use the two above toy objects and test the whole message in the two cases. 
Is that correct ?

wdevazelhes commented:
> Looks nice overall. Please remove `sklearn/linear_model/tests/__pycache__/tmp1ax1d_im`.

Done

Commit:
Merge branch 'master' into tests_sparse

Commit:
Change ValueError into TypeError for QuantileTransformer

Commit:
Modify also the test

Commit:
Avoid to reiterate the distance matrix of the distance matrix etc

Commit:
Put the right error for error message in approximate nearest neighbors

Commit:
Also update the test with TypeError

wdevazelhes commented on sklearn/utils/estimator_checks.py:
```diff
@@ -2416,6 +2417,69 @@ def check_classifiers_regression_target(name, estimator_orig):
+    centers = 2 if tags["binary_only"] else None
+    X, y = make_blobs(random_state=rng, cluster_std=0.5, centers=centers)
+    # for put some points to zero to have a little bit of sparsity
+    if tags["requires_positive_X"]:
```
This could be updated with `enforce_estimator_tags_X` as soon as #14705 is merged

wdevazelhes commented on sklearn/utils/estimator_checks.py:
```diff
@@ -2416,6 +2417,69 @@ def check_classifiers_regression_target(name, estimator_orig):
+    if tags["requires_positive_X"]:
+        X -= np.min(X) - 1.
+    X_csr = sparse.csr_matrix(X)
+    if tags["multioutput"] or tags["multioutput_only"]:
```
Should this use `enforce_estimator_tags_y` ? The latter does not transform the input in the case of "multioutput" (only "multioutput_only").

wdevazelhes commented on sklearn/utils/estimator_checks.py:
```diff
@@ -2416,6 +2417,69 @@ def check_classifiers_regression_target(name, estimator_orig):
+                estimator.set_params(with_mean=False)
+                estimator_sp.set_params(with_mean=False)
+            if name in ['RobustScaler']:
+                # XXX naming is not consistent (with_mean vs with_centering) :(
```
Should I open an issue for that ?

agramfort replied:
yes +1 so at we can discuss it. Once the issue is opened you can remove this comment.

wdevazelhes replied:
Done: #14810

wdevazelhes commented:
Except my last comments, I guess this PR should be ready to merge

Commit:
Remove comment since the issue is opened

wdevazelhes committed:
merge with master

wdevazelhes committed:
fix pep8 errors

wdevazelhes committed:
Put TypeError instead of ValueError in neighbors tree

wdevazelhes committed:
update QuantileTransformer with TypeError when sparse not accepted

wdevazelhes commented:
I just realized that the test I added in `test_estimator_checks` to address https://github.com/scikit-learn/scikit-learn/pull/13246#discussion_r309167624 will in fact not work (i.e. not test the actuall added lines in `estimator_checks`, hence maybe the coverage pb), because the right exception will be returned by `check_estimator_sparse_data` (the test just before `check_estimator_sparse_dense`), hence skipping `check_estimator_sparse_dense`). So I think I should actually merge these two tests (because a lot of what they do is similar anyway), and also it made me realize that I don't test (in `test_check_estimator`) what happens if the sparse version and the dense version are different, which I should do too. 
I'll do that soon

wdevazelhes committed:
Remove the test of exceptions in check_estimator_sparse_dense

wdevazelhes committed:
Merge branch 'master' into tests_sparse

wdevazelhes committed:
- Test that check_estimator_dense_sparse indeed checks equality
- Make the fake estimators inherit from LogisticRegression so
  that they pass all checks except the one they are made for

wdevazelhes committed:
STY: fix pep8 errors

wdevazelhes committed:
Use a BaseBadClassifier to not raise the NotFittedError since NoSparseClassifier now inherits from LogisticRegression (which raises the right error)

wdevazelhes committed:
no reason to remove this decorator, must be a mistake in merge or sth

wdevazelhes committed:
be clearer about NoSparseClassifier

wdevazelhes commented on sklearn/utils/estimator_checks.py:
```diff
@@ -2565,6 +2567,60 @@ def check_classifiers_regression_target(name, estimator_orig):
+                probs_sp = estimator.predict_proba(X_sp_converted)
+                assert probs_sp.shape == (X.shape[0], len(np.unique(y)))
+                assert_allclose(probs, probs_sp)
+        except (ValueError, TypeError):
```
In the end I decided to keep two different tests, but to remove the part where we check the error message if sparse inputs is not supported here, since it's the part already tested by `check_estimator_sparse_data`
What do you think ?

ogrisel replied:
This is a problem because `ValueError` or `TypeError` could be raised by the code in the checks, after you call `fit`. For instance when you call `predict_proba` in the check.

So please instead do something like:

```

        try:
            estimator_sp.fit(X_sp_converted, y)
            estimator.fit(X_converted, y)
         except (ValueError, TypeError):
            # This is the case where estimators don't support sparse inputs.
            # We don't want to have this case interfere  with our test of equality:
            # it is already tested by check_estimator_sparse_data
            continue
        # put the rest of the code specific to check_estimator_sparse_dense
        # here
        ...
```

wdevazelhes replied:
I agree, done, I used the try/except/else statement, so that only estimators that raised no exception (so fitted in the "try"), will be asked to `predict`.

wdevazelhes commented on sklearn/utils/tests/test_estimator_checks.py:
```diff
@@ -151,16 +151,63 @@ def fit(self, X, y):
-class NoSparseClassifier(BaseBadClassifier):
-    def fit(self, X, y):
-        X, y = check_X_y(X, y, accept_sparse=['csr', 'csc'])
+class NoSparseClassifierException(LogisticRegression):
```
I made the fake estimators inherit from `LogisticRegression` so that they pass all the `check_estimator` checks, (except the one except the one we want to test like `check_estimator_sparse_data` for NoSparseClassifier, `NoSparseClassifierTypeError` or `NoSparseClassifierValueError`, or `check_estimator_sparse_dense` for `SparseDenseDifferentPredict` or `SparseDenseDifferentPredictProba`)

wdevazelhes replied:
(Note that there was also `check_estimator_unfitted` that used `NoSparseClassifier`, but I made it use a plain `BaseBadClassifier` instead so that it indeed uses a Classifier that doesn't raise an error message when not fitted)

ogrisel replied:
I think it would be enough to derive from `BaseBadClassifier`. This is only meant to check that the estimator checks that checks the error message is informative right? Let's keep it as simple as possible.

ogrisel replied:
Also the name `NoSparseClassifierException` is weird because one my except it to be a exception class (a subclass of `Exception`).

Instead I would name it something like `BadFitExceptionSparseClassifier` or something.

wdevazelhes replied:
Originally I inherited from `LogisticRegression` so that no error, other than the one I wanted, was raised (I just wanted `Exception("Nonsensical Error")` to be raised) . But I agree it's a bit confusing, so instead I falled back to inheriting from `BaseBadClassifier` as you said and just ensured to do the necessary checks so that no unwanted error is thrown (see  https://github.com/scikit-learn/scikit-learn/pull/13246/commits/70e9d2c9bc103f4bc734789e376706c7cabb92f3)

wdevazelhes replied:
I also changed the name to `BadFitExceptionSparseClassifier `

wdevazelhes commented:
I made a few modifications (see comments above), and I think the PR is ready to merge, what do you think ?

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -2565,6 +2567,60 @@ def check_classifiers_regression_target(name, estimator_orig):
+            if hasattr(estimator, "predict"):
+                pred = estimator.predict(X_converted)
+                pred_sp = estimator_sp.predict(X_sp_converted)
+                if not (isinstance(estimator, BaseSGD) or  # SGD is too random
```
I wish we didn't need to make these exceptions by name. But it's not clear which exceptions deserve having tags named for them. Now that we have decomposed check_estimator, I would like to try implementing these exceptions in test_common rather than here... But maybe that belongs in another pr

wdevazelhes replied:
I am not really familiar with `test_common` but it seems to be the place to run tests with particular values for estimators parameters, is that it ? So do you mean that we should run a test for equivalence between sparse and dense for `StandardScaler` in `test_common` rather than in `sklearn.preprocessing.tests.test_data` for instance ?

ogrisel replied:
Estimator checks are used both internally in scikit-learn to ensure that all estimators have a consistent behaviors and error messages for edge cases. But they are also used by third party libraries to check that their estimators are compatible with the scikit-learn way of doing things. Because those checks can be used by third party libraries, we would rather like them not depend on specific estimators in scikit-learn. In that respect the `isinstance(estimator, BaseSGD)` is a red flag.

If some estimators should skip a particular test for good reason, we usually tag them with a dedicated estimator tag:

https://scikit-learn.org/stable/developers/develop.html#estimator-tags

See in particular `_xfail_checks`.

The scikit-learn internal thing that run the estimator checks on scikit-learn estimators is the `test_common`, see: `sklearn/tests/test_common.py`.

ogrisel replied:
Still I find it intriguing that the randomness of SGD has an impact on this check. We call `set_random_state(estimator)` and `set_random_state(estimator_sp)` so this is not a good reason in itself. Maybe there is a real bug with this estimator.

wdevazelhes committed:
Use the undeprecated version of _pairwise_estimator_convert

wdevazelhes commented on sklearn/utils/estimator_checks.py:
```diff
@@ -2565,6 +2567,60 @@ def check_classifiers_regression_target(name, estimator_orig):
+    for sparse_format in ['csr', 'csc', 'dok', 'lil', 'coo', 'dia', 'bsr']:
+        X_sp = X_csr.asformat(sparse_format)
+        if name in ['Scaler', 'StandardScaler']:
+            estimator.set_params(with_mean=False)
```
I agree, I'll remove them here
I checked and it seems that these variants are in fact actually already tested here `sklearn.preprocessing.tests.test_scaler_without_centering` (by the way I didn't find any `Scaler` class in scikit-learn so I guess it was just really about `StandardScaler`), so I guess I don't need to implement a new test ?

ogrisel replied:
This comment still needs to be addressed.

> (by the way I didn't find any Scaler class in scikit-learn so I guess it was just really about StandardScaler), 

And `RobustScaler`.

> so I guess I don't need to implement a new test ?

If `RobustScaler` is also tested, then no need for a new test. I haven't checked.

wdevazelhes replied:
Yes, `RobustScaler`is tested too here: https://github.com/scikit-learn/scikit-learn/blob/1c116807ce1e925f73c87c34e586520d455682d6/sklearn/preprocessing/tests/test_data.py#L1123

So I removed the lines with the "ifs" here in `estimator_checks`, and added in `StandardScaler` and `RobustScaler` the `xfail` estimator tag for this `check_estimator_dense_sparse` test (I explain in the `xfail` message that I exclude this test because it's not possible for the default parameters, but the test is made in `test_scaler_without_centering/test_robust_scaler_equivalence_dense_sparse`) (see commit https://github.com/scikit-learn/scikit-learn/pull/13246/commits/e32e6aa87c51f47fd50d62536bb9287d7966c8c5)


PS: I noticed however that `StandardScaler`'s test is more intensive (tests `transform`, `inverse_transform`, etc, when the one for `RobustScaler` only tests the `mean_` attribute) https://github.com/scikit-learn/scikit-learn/blob/1c116807ce1e925f73c87c34e586520d455682d6/sklearn/preprocessing/tests/test_data.py#L785
I thought I could make `RobustScaler`'s test as robust as `StandardScaler`'s one, but I don't know if it's the purpose of this PR maybe it just adds more to review ?

wdevazelhes commented on sklearn/utils/estimator_checks.py:
```diff
@@ -2565,6 +2567,60 @@ def check_classifiers_regression_target(name, estimator_orig):
+                pred = estimator.predict(X_converted)
+                pred_sp = estimator_sp.predict(X_sp_converted)
+                if not (isinstance(estimator, BaseSGD) or  # SGD is too random
+                        getattr(estimator, 'kernel', None) == 'precomputed' or
```
Removing these lines, it works for pairwise estimators (maybe because we use `_pairwise_estimator_convert_X` before ?)
But it does not work for `BaseSGD`, which seems to be non deterministic, I tried to look for an issue about that but it seems that the flag `not_deterministic` is not given to `BaseSGD`, maybe it should ?

ogrisel replied:
`BaseSGD` should be deterministic (assuming you fix the `random_state`). I believe `not_deterministic` is better suited for other uncontrollable sources of non-determinism (e.g. slight numerical variation caused by different ordering of arithmetic operations caused by thread-based parallelism that causes ties in cluster assignment to be broken in different ways when running k-means clustering for instance).

adrinjalali commented:
Moving this to 0.24, but happy to have it in 0.23 if it finishes on time.

cmarmo commented:
HI @wdevazelhes , are you still interested in finishing this? Would you be available to address comments and fix conflicts? Thanks for your work so far!

wdevazelhes commented:
Hi @cmarmo, sorry for letting this like this, I’ll try to give it a shot next week !

wdevazelhes committed:
merge tentative

wdevazelhes committed:
remove remaining from merge

wdevazelhes committed:
fix merge

wdevazelhes committed:
fixes

wdevazelhes committed:
pep8 fix

wdevazelhes committed:
replace _safe_tags by _get_tags

wdevazelhes committed:
put again the original code that I accidently removed in the merge

wdevazelhes committed:
fix indentation

wdevazelhes committed:
the block should be in fact removed

wdevazelhes committed:
stochastic gradient not found, replacing by private module

wdevazelhes committed:
add strict mode

wdevazelhes committed:
use instance not class

wdevazelhes committed:
fix tests

wdevazelhes commented:
I didn't have the time to finish yet, but will try to continue this week!

wdevazelhes commented on sklearn/utils/estimator_checks.py:
```diff
@@ -813,6 +797,24 @@ def check_estimator_sparse_data(name, estimator_orig, strict_mode=True):
                 else:
                     expected_probs_shape = (X.shape[0], 4)
                 assert probs.shape == expected_probs_shape
+        except (TypeError, ValueError) as e:
```
I put back the old try/except here (instead of the newly introduced `with raises(...`), because I want to test first that if a `TypeError` or `ValueError` has been thrown, the right error msg is thrown, and then that if it's not this type of error but some general `Exception` that has been thrown, a `TypeError` or `ValueError` should be thrown, and I didn't manage to do this hierarchical testing with `with raises(...`

wdevazelhes committed:
remove message since referenced pr is closed

wdevazelhes committed:
remove pyorig file

wdevazelhes committed:
use enforce_estimator_tags

ogrisel commented on sklearn/utils/estimator_checks.py:
```diff
@@ -813,6 +797,24 @@ def check_estimator_sparse_data(name, estimator_orig, strict_mode=True):
+            msg = ("Estimator %s doesn't seem to fail gracefully on "
+                   "sparse data: it should raise a TypeError or a ValueError "
+                   "if sparse input is explicitly not supported.")
+            raise AssertionError(msg % (name))
```
nitpick: could you please use f-strings for all those messages instead of %-formatting?

wdevazelhes replied:
Sure, done

ogrisel commented on sklearn/utils/tests/test_estimator_checks.py:
```diff
@@ -173,16 +173,63 @@ def fit(self, X, y):
+        return super(NoSparseClassifierException, self).fit(X, y)
+
+
+class NoSparseClassifierTypeError(LogisticRegression):
```
Similar comment as above for this class the the following.

ogrisel reviewed (commented):
Another pass of comments:

wdevazelhes committed:
use fstrings

wdevazelhes committed:
put a standard random seed

wdevazelhes commented:
> So this mean that this check would pass for all scikit-learn estimators with a canonical random seed such as 0 or 42? If so please change the seed to 0 or 42 because 52 makes us think that it's a special seed.

Yes it would, you're right, I just updated the random seed to 42

wdevazelhes committed:
rename mock classes with better names

wdevazelhes commented:
> Also the name NoSparseClassifierException is weird because one my except it to be a exception class (a subclass of Exception).
> Instead I would name it something like BadFitExceptionSparseClassifier or something.

I agree, done

wdevazelhes commented:
(sorry for the delay, I'll continue to address the comments probably later this week or this week end)

wdevazelhes committed:
address https://github.com/scikit-learn/scikit-learn/pull/13246/files/eb854c04065984657f13c1aeb1e99c6e50b2ab31\#discussion_r512857180

wdevazelhes committed:
Allow sparsity in AffinityPropagation and fix error in check that was previously silenced by my wrong try/except

wdevazelhes committed:
Fix pb that appeared due to the previous fix:

- NotInvariantSampleOrder was missing the multioutput tag, and didn't work on sparse inputs
- NotInvariantPredict accepted sparse inputs at fit time but not at predict time

wdevazelhes committed:
Remove specific centering for StandardScaler and RobustScaler in estimator_checks, and add an xfail flag fot them

wdevazelhes committed:
fix indent

wdevazelhes committed:
Remove if precomputed metric or kernel, see comment https://github.com/scikit-learn/scikit-learn/pull/13246\#discussion_r332027433

wdevazelhes committed:
try to not exclude SGD to see if it still works

agramfort commented:
@wdevazelhes you have a conflict

let me know if you need help here

wdevazelhes committed:
Merge branch 'master' into tests_sparse

wdevazelhes commented:
Thanks @agramfort , for the conflict, there was only one line conflicting, for which I just added the "accept_sparse=True" flag like this: https://github.com/scikit-learn/scikit-learn/pull/13246/commits/d7587e1318cb3b9d4795f1b501286160a7f75185

I still need to address some comments, plus I currently have ``SGDRegressor`` that don't pass the test, that I'm investigating: I removed the `if not (isinstance(estimator, BaseSGD)` (see comment https://github.com/scikit-learn/scikit-learn/pull/13246#discussion_r512852837) so that the test also applies to SGD estimators, and they all pass, but ``SGDRegressor`` didn't pass... It's maybe due to the sparse intercept decay here ? https://github.com/scikit-learn/scikit-learn/blob/35320845c24ee58a21a99a2df044084e0e65f3a4/sklearn/linear_model/_base.py#L46, that is called here: https://github.com/scikit-learn/scikit-learn/blob/35320845c24ee58a21a99a2df044084e0e65f3a4/sklearn/linear_model/_stochastic_gradient.py#L1291, because if I set  ``SPARSE_INTERCEPT_DECAY = 1`` (i.e. same as the non sparse intercept decay), this time the test passes. I'll look into this

Commit:
Use BaseBadClassifier instead of LogisticRegression

wdevazelhes commented:
Regarding the tests that fail when applied on children of `BaseSGD`, I'm thinking to exclude those from passing through `check_estimator_sparse_dense` tests using `_xfail_checks`, and ensure there is a test in `linear_model/tests/test_sgd.py` that:
- sets the intercept to the same value btw sparse/nonsparse and 
- checks on a toy example that the result is the same
What do you think ?

wdevazelhes committed:
fix typo

wdevazelhes committed:
Add checks for BaseSGD that sets the intercept decay to the same value between sparse and dense

wdevazelhes committed:
remove unused import

agramfort commented:
AFAIK the learning rate for intercept in different in sparse case for SGD

wdevazelhes committed:
try with a smaller intercept decay to be more stable ?

wdevazelhes committed:
fix code to set the intercept decay to a smaller value for both

wdevazelhes committed:
fix code to set the intercept decay to a smaller value for both

wdevazelhes committed:
ensure that the test doesn't change the intercept decay

wdevazelhes committed:
remove blank line

wdevazelhes committed:
Be more specific by excluding only estimators that fail (with the justification that they use a different value for the intercept decay btw sparse and dense

wdevazelhes committed:
remove unused imports

wdevazelhes commented:
> AFAIK the learning rate for intercept is different in sparse case for SGD

Indeed, so in the end I excluded from the common tests specifically the two tests that failed because of this, namely `RegressorSGD` and  `PassiveAgressiveRegressor`, but I made them a specific test, that checks that if the intercept decay was the same, they would return the same result. See this commit: https://github.com/scikit-learn/scikit-learn/pull/13246/commits/138b2b9252fadc9941d94c8010c39fbf132f4e96

wdevazelhes commented:
There still seems to be a problem for the `Linux32 py36_ubuntu_atlas_32_bits` build, regarding `NuSVR`: the results are almost the same, but not within the allowed tolerance. Maybe this is related to https://github.com/scikit-learn/scikit-learn/pull/16530 ? If this precision is acceptable, I can maybe do the same strategy as in the previous comment, i.e. exclude `NuSVR` from the common tests, but create a specific test with a higher tolerance just for `NuSVR` ?

agramfort commented:
I would try to use a dataset with better conditioning / easier to find
solution such that numerical issues are less likely to happen

>

wdevazelhes commented:
> I would try to use a dataset with better conditioning / easier to find solution such that numerical issues are less likely to happen

Thanks @agramfort , indeed, all the estimators that failed are regressions, and I realized I tried to fit them on a classification task (built with `make_blobs`), so indeed the problem was not easy for them... 
I'll update the test using a regression task ( with `make_regression`) for regressors (found using `is_regressor`), it should solve the pb, and maybe I'll even can remove the `_xfail_checks` tags that I added

wdevazelhes commented:
I'm managing to have a well conditioned pb for regressions in general, however ensembling of regression still give slightly different results, which seems even more related to this pb https://github.com/scikit-learn/scikit-learn/pull/16530#issuecomment-608336523, so I'll try to understand if I can have an even better conditioned problem, and also explore how this pb has been solved before (for instance, it seems that the ensembling methods can be tested with no pb by using, instead of the default regressions which uses libsvm, a more stable implementation like `LinearSVC`)

agramfort commented:
you need help?

>

wdevazelhes committed:
Create simple regression pb for regressors

wdevazelhes commented:
I created a simple regression pb that is well conditioned, and small (i.e. a few dims and a few samples) so that there are less risks for numerical problems (e.g. there will be fewer computations to do at each iteration of gradient), it passed the tests on my machine, let's see if it passes them here

wdevazelhes commented:
> you need help?

Thanks a lot, I think if the tests passes now, it should be almost finished, there's just one thing I would still want to add for the test to be complete: instead of checking just fit and predict/predict_proba, I should also test transform, and maybe even iterate through all predictions functions like `for func in ["predict", "predict_proba", "transform"]` etc, like it can be done in other tests

(the most important is to check the `transform` I think, which is not done right now)

wdevazelhes commented:
Tests still don't pass on the  `Linux32 py36_ubuntu_atlas_32bit` and `Linux pylatest_pip_openblas_pandas` platforms :p @agramfort I'll try to search, but maybe there is a way to make the toy dataset stable even for 32 bits architectures ? (Maybe by creating a dataset with numbers with a few significative numbers like `0.600000` or sth like this?)

wdevazelhes committed:
round the dataset to limit numerical problems

wdevazelhes commented:
I have tried to round the dataset to 2 decimal digits, let's see if it works

wdevazelhes commented:
it still doesnt pass, but the difference is much smaller, so I'll try instead of rounding to 2 decimals, to round to 1 significant digit

wdevazelhes committed:
put less significative numbers

wdevazelhes committed:
try with float16 precision

wdevazelhes commented:
Tests are passing, using a dataset made of float16 numbers (converted back to float32). I am not sure to understand why and if it's by chance, but I think it can be acceptable ? i.e. I didn't set any random seed here, and didn't tweak the coeffs in the linear regression (I just took some reasonable values rather than taking the risk to have almost zero coeffs if I had set them randomly)

I'll go on with testing the ``transform`` function too

agramfort commented on sklearn/utils/estimator_checks.py:
```diff
@@ -2939,6 +2944,65 @@ def check_classifiers_regression_target(name, estimator_orig,
+        X = X[~np.all(X == 0, axis=1)]  # we remove null rows
+        assert len(np.where(X == 0)[0]) > 0  # we should have sparsity
+        assert X.shape[0] > 5  # we mustn't have very few samples
+        size_X = X.shape[0]
```
```suggestion
        n_samples = X.shape[0]
```

agramfort commented on sklearn/utils/estimator_checks.py:
```diff
@@ -2939,6 +2944,65 @@ def check_classifiers_regression_target(name, estimator_orig,
+        assert X.shape[0] > 5  # we mustn't have very few samples
+        size_X = X.shape[0]
+        y = X.dot(np.array([0.1, -0.2])) + 0.01 * rng.randn(size_X)
+        y = y.astype(np.float16).astype(np.float)
```
add a note why you added this.

wdevazelhes replied:
done here: https://github.com/scikit-learn/scikit-learn/pull/13246/commits/70d6e889d5c1d2ba942601f7afa3f41537db0564, let me know what you think about it (I'm not completely sure of it):

```
        # we build float64 precision numbers from low precision ones:
        # this way, fewer numbers get dropped by the
        # machine precision in multiplications/additions, so if those
        # are done in a different order btw sparse/dense, the results
        # are more likely to still be the same
```

agramfort commented on sklearn/utils/estimator_checks.py:
```diff
@@ -2939,6 +2944,65 @@ def check_classifiers_regression_target(name, estimator_orig,
+        tags = estimator_orig._get_tags()
+        centers = 2 if tags["binary_only"] else None
+        X, y = make_blobs(random_state=rng, cluster_std=0.5, centers=centers)
+    # for put some points to zero to have a little bit of sparsity
```
```suggestion
    # set some points to zero to have a little bit of sparsity
```

agramfort replied:
But I don't see what this refers to

wdevazelhes replied:
you're right, I think it's a remainder from some previous code that I removed
I removed the comment

agramfort reviewed (commented):
besides my nitpicks LGTM

nice work @wdevazelhes !

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -2939,6 +2944,65 @@ def check_classifiers_regression_target(name, estimator_orig,
+                probs_sp = estimator_sp.predict_proba(X_sp_converted)
+                assert probs_sp.shape == (X.shape[0], len(np.unique(y)))
+                assert_allclose(probs, probs_sp)
+
```
worth checking decision_function too?

wdevazelhes replied:
I agree, done

wdevazelhes replied:
I agree, done

jnothman reviewed (commented):
This is a nice extra check.

I wonder whether it's appropriate to test non-canonical sparse representations, e.g. where a CSR matrix duplicates an index (which means that the values at that index should be summed).

wdevazelhes commented:
Thanks for the review @agramfort and @jnothman ! I'll address them this week

wdevazelhes commented:
(sorry I didn't manage to do it last week, I'll try this week)

wdevazelhes committed:
Update sklearn/utils/estimator_checks.py

Co-authored-by: Alexandre Gramfort <alexandre.gramfort@m4x.org>

wdevazelhes committed:
refactor X_size into n_samples

wdevazelhes committed:
Update sklearn/utils/estimator_checks.py

Co-authored-by: Alexandre Gramfort <alexandre.gramfort@m4x.org>

wdevazelhes committed:
remove comment

wdevazelhes committed:
add comment about the precision of numbers

wdevazelhes committed:
add other prediction functions

wdevazelhes committed:
Remove Scalers tests

wdevazelhes commented:
I just removed the tests of scalers in `check_estimator_sparse_data`, because 
- they are already tested elsewhere, and
- explicits exceptions by estimator name should be avoided in common tests 

(see https://github.com/scikit-learn/scikit-learn/pull/13246#discussion_r330438412))

wdevazelhes committed:
separate scoring function and prediction/transform functions

wdevazelhes committed:
score_samples dont have y

wdevazelhes committed:
test more methods in check_estimator_sparse_data

wdevazelhes committed:
the try except should only be on the fit, we don't want assertion errors to raise them

wdevazelhes committed:
subsequent tests should be done only on estimators that pass the 'fit' step

wdevazelhes committed:
make shape testing work with outlier detector

wdevazelhes commented:
I did this: https://github.com/scikit-learn/scikit-learn/pull/13246/commits/d066b328dfdd42104ac2f0e29f94a18860cbba71
because in the case of outlier detection for instance, an estimator can predict +1 or -1 (outlier/not outlier), even when the provided y is multiclass: (they will actually ignore the y because they are unsupervised). So I think there's no way to know in advance what is the dimension of predictions, so I only check if there are `n_sample` predictions. (and of course the later test `check_estimator_sparse_dense` will still check that the dense predictions and sparse predictions match, so I think it's robust enough)

wdevazelhes commented:
The new tests (for other functions like `transform` etc) shed light on some bugs, some of which can be solved without too much trouble (for isntance some code tries to select elements of dia sparse matrix with indexing which doesn't work), some seem more difficult (numerical uncertainties), I'll address those after the holidays

wdevazelhes committed:
Fix AdditiveChi2Sampler

wdevazelhes committed:
fix IncrementalPCA: accept to transform some sparse inputs with index not subscriptable

wdevazelhes commented:
This issue: https://github.com/scikit-learn/scikit-learn/issues/17554, is causing the following test not to pass: ``test_estimators[PolynomialFeatures()-check_estimator_sparse_data(strict_mode=True)]``. I'll comment on that issue to see if I can fix it.

wdevazelhes committed:
Fix problems with linear_models

wdevazelhes committed:
Simplify data to remove numerical uncertainties

wdevazelhes commented on sklearn/utils/estimator_checks.py:
```diff
@@ -2974,7 +2974,9 @@ def check_estimator_sparse_dense(name, estimator_orig,
         tags = estimator_orig._get_tags()
         centers = 2 if tags["binary_only"] else None
-        X, y = make_blobs(random_state=rng, cluster_std=0.5, centers=centers)
+        X, y = make_blobs(n_samples=10, n_features=2, random_state=rng,
```
I simplified the data here similarly to before, because there were some numerical uncertainties with the Nystrom method (particularly with almost zero elements: the relative error between two numbers can be quite high but both number can be of order 1e-15 for instance, so they are clearly supposed to be zeros). With a simpler dataset that will lead to less operations (because less samples, less features, and less decimals), the uncertainties disappeared on my machine

cmarmo commented:
Hi @wdevazelhes , do you mind synchronizing with upstream? Checks are failing looking for the master branch. Thanks!

wdevazelhes committed:
Merge branch 'master' into tests_sparse

wdevazelhes commented:
Hi @cmarmo , thanks, I just merged with master. 
I also added a few things in the TODO that I need to finish before putting the PR in MRG

wdevazelhes committed:
Refactor dataset use

wdevazelhes committed:
use the more extensive _generate_sparse_matrix

wdevazelhes commented on sklearn/utils/estimator_checks.py:
```diff
@@ -780,30 +780,16 @@ def check_estimator_sparse_data(name, estimator_orig, strict_mode=True):
                     assert pred.shape == (X.shape[0], 1)
+                elif tags['multioutput']:
+                    assert (pred.shape == (X.shape[0], 1) or
+                            pred.shape == (X.shape[0],))
```
You're right, here I test only with a mono-output (binary) y, so we know the array will be 1D, so I removed these lines. By the way, let me know if you think we should test multioutput estimators with multioutput datasets too. I just devised the test with a binary y for classifiers and a continuous y for regressors, because I just wanted to be able to run the estimator, on sparse and dense inputs and see if they have the same behaviour. But I could do more intensive tests, using several multiclass/multioutput datasets ?

wdevazelhes committed:
fix ortho

wdevazelhes committed:
add changelog

wdevazelhes committed:
use float because np.float is deprecated

wdevazelhes committed:
merge with master

adrinjalali commented:
removing the milestone for now.

haiatn commented:
Is PR ready for review?

agramfort commented:
@haiatn this one would deserve a big refresher to be now considered for merge. I don't know how much we should invest on this one at this point.

haiatn commented:
> @haiatn this one would deserve a big refresher to be now considered for merge. I don't know how much we should invest on this one at this point.

I do like the idea of testing this, but if you think so, maybe we should close this and leave the issue open for whoever that would like to start fresh

agramfort commented:
you can see how difficult it would be to rebase this PR on current main
branch. Feel free to push directly to this PR
or open a new one.

> Message ID: ***@***.***>
>

adrinjalali commented:
I think this can be closed, the original issue is still open and relevant
